File: ios/MusicSearch/Networking/MusicModel.swift
import Foundation

/// 核心音乐业务 API（对应 Android MusicModel）
enum MusicModel {

    private static let bangMusicUrl = "http://kbangserver.kuwo.cn/ksong.s"
    private static let playListUrl = "https://mobilist.kuwo.cn/list.s"
    private static let searchUrl = "https://search.kuwo.cn/r.s"
    private static let musicPlayUrl = "http://nmobi.kuwo.cn/mobi.s?f=kuwo&q="
    private static let musicLyricUrl = "http://m.kuwo.cn/newh5/singles/songinfoandlrc"
    private static let musicDetailUrl = "https://musicpay.kuwo.cn/music.pay"

    // MARK: - 榜单音乐
    static func getBangMusic(id: String, page: Int, size: String) async throws -> [MusicBean] {
        let params = ["from": "pc", "type": "bang", "id": id, "pn": String(page), "rn": size]
        let resp = try await HttpRequest.request(url: bangMusicUrl, params: params)
        return parseMusicList(resp, key: "musiclist")
    }

    // MARK: - 歌单音乐
    static func getPlayListMusic(id: String, page: Int, size: String) async throws -> [MusicBean] {
        let params = ["type": "songlist", "id": id, "pn": String(page), "rn": size]
        let resp = try await HttpRequest.request(url: playListUrl, params: params)
        // 结构: { data: { musiclist: [...] } }
        guard let data = resp.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = root["data"] as? [String: Any] else { return [] }
        let arr = dataObj["musiclist"] as? [[String: Any]] ?? []
        return parseMusicArray(arr)
    }

    // MARK: - 搜索音乐
    static func getSearchMusic(name: String, page: Int, size: String) async throws -> [MusicBean] {
        let params: [String: String] = [
            "client": "kt", "all": name, "vipver": "1", "ft": "music",
            "encoding": "utf8", "rformat": "json", "mobi": "1",
            "pn": String(page), "rn": size
        ]
        let resp = try await HttpRequest.request(url: searchUrl, params: params)
        guard let data = resp.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let arr = root["abslist"] as? [[String: Any]] else { return [] }
        return arr.compactMap { item in
            let rid = item["DC_TARGETID"] as? String ?? ""
            let n = item["NAME"] as? String ?? ""
            let a = item["ARTIST"] as? String ?? ""
            var img = item["web_albumpic_short"] as? String ?? ""
            if !img.isEmpty && !img.hasPrefix("http") {
                img = "https://img4.kuwo.cn/star/albumcover/" + img
            }
            if img.isEmpty || img.count < 5 {
                img = "https://img1.kuwo.cn/star/albumcover/default.jpg"
            }
            return MusicBean(id: rid, name: n, artist: a, albumIcon: img, like: false)
        }
    }

    // MARK: - 播放地址
    static func getMusicPlayInfo(musicId: String, bitrate: String) async throws -> String {
        let api = "user=0&android_id=0&prod=kwplayerhd_ar_4.3.0.8&corp=kuwo&vipver=4.3.0.8&source=kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk&notrace=0&type=convert_url2&br=\(bitrate)&format=flac|mp3|aac&sig=0&priority=bitrate&loginUid=0&network=WIFI&loginSid=0&mode=down&rid=\(musicId)"
        let encrypted = EncryptMusicToken.encrypt(api)
        let url = musicPlayUrl + encrypted
        return try await HttpRequest.request(url: url)
    }

    // MARK: - 歌词
    static func getMusicLyric(musicId: String) async throws -> String {
        let params = ["musicId": musicId]
        // 该接口需要 Cookie
        return try await HttpRequest.requestWithCookie(url: musicLyricUrl, params: params,
            cookie: "kw_token= music")
    }

    // MARK: - 音质详情
    static func getMusicDetail(musicId: String) async throws -> [(bitrate: String, size: String)] {
        let params = ["op": "query", "action": "play", "ids": musicId]
        let resp = try await HttpRequest.request(url: musicDetailUrl, params: params)
        guard let data = resp.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let songs = root["songs"] as? [[String: Any]],
              let first = songs.first,
              let minfo = first["MINFO"] as? String else { return [] }
        // MINFO 格式: "bitrate=2000|format=flac|size=12345;bitrate=320|format=mp3|size=6789;..."
        return minfo.split(separator: ";").compactMap { segment in
            let parts = segment.split(separator: "|")
            var br = "", sz = ""
            for p in parts {
                let kv = p.split(separator: "=")
                if kv.count == 2 {
                    if kv[0] == "bitrate" { br = String(kv[1]) }
                    if kv[0] == "size" { sz = String(kv[1]) }
                }
            }
            return br.isEmpty ? nil : (bitrate: br, size: sz)
        }.sorted { Int($0.bitrate) ?? 0 > Int($1.bitrate) ?? 0 }
    }

    // MARK: - 下载地址
    static func getMusicDownloadUrl(musicId: String, bitrate: String) async throws -> String? {
        let api = "user=0&android_id=0&prod=kwplayerhd_ar_4.3.0.8&corp=kuwo&vipver=4.3.0.8&source=kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk&p2p=1&notrace=0&type=convert_url2&br=\(bitrate)&format=flac|mp3|aac&rid=\(musicId)&priority=bitrate&loginUid=0&network=WIFI&loginSid=0&mode=down"
        let encrypted = EncryptMusicToken.encrypt(api)
        let url = musicPlayUrl + encrypted
        let resp = try await HttpRequest.request(url: url)
        return MySubString.subString(resp, start: "url=", end: "sig")
    }

    // MARK: - 从播放地址响应中提取 url
    static func extractPlayUrl(_ resp: String) -> String? {
        return MySubString.subString(resp, start: "url=", end: "sig")
    }

    // MARK: - 通用解析
    private static func parseMusicList(_ json: String, key: String) -> [MusicBean] {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let arr = root[key] as? [[String: Any]] else { return [] }
        return parseMusicArray(arr)
    }

    private static func parseMusicArray(_ arr: [[String: Any]]) -> [MusicBean] {
        return arr.compactMap { item in
            let id = String(item["rid"] as? Int ?? 0)
            guard id != "0" || item["rid"] != nil else { return nil }
            let name = item["name"] as? String ?? ""
            let artist = item["artist"] as? String ?? ""
            var img = item["albumpic"] as? String ?? item["pic"] as? String ?? item["img"] as? String ?? item["album_img"] as? String ?? item["album_pic"] as? String ?? ""
            if !img.isEmpty && img.count >= 5 && !img.hasPrefix("http") {
                img = "https://img4.kuwo.cn/star/albumcover/" + img
            }
            if img.isEmpty || img.count < 5 {
                img = "https://img1.kuwo.cn/star/albumcover/default.jpg"
            }
            return MusicBean(id: id, name: name, artist: artist, albumIcon: img, like: false)
        }
    }
}

File: app/src/main/java/com/linfeng/music/model/MusicModel.java
package com.linfeng.music.model;

import com.linfeng.music.utils.EncryptMusicToken;
import com.linfeng.music.utils.HttpRequest;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import io.reactivex.rxjava3.core.Single;

public class MusicModel {
    /**
     * 获取榜单音乐详情的API
     */
    private static final String BangMusicUrl = "http://kbangserver.kuwo.cn/ksong.s";
    /**
     * 获取歌单音乐详情的API
     */
    private static final String PlayListUrl = "https://mobilist.kuwo.cn/list.s";
    /**
     * 搜索音乐详情的API
     */
    private static final String SearchUrl = "https://search.kuwo.cn/r.s";
    /**
     * 获取音乐url的API
     */
    private static final String MusicPlayUrl = "http://nmobi.kuwo.cn/mobi.s?f=kuwo&q=";
    /**
     * 获取音乐歌词的API
     */
    private static final String MusicLyricUrl = "http://m.kuwo.cn/newh5/singles/songinfoandlrc";
    /**
     * 获取音乐的支持音质详情
     */
    private static final String MusicDetailUrl = "https://musicpay.kuwo.cn/music.pay";



    /**
     * 获取榜单音乐的List数据
     *
     * @param id   榜单id
     * @param page 页数
     * @param size 长度
     * @return 未解析的数据
     */
    public Single<String> getBangMusic(String id, int page, String size) {
        // 设置请求参数
        Map<String, String> params = new HashMap<>();
        params.put("from", "pc");
        params.put("type", "bang");
        params.put("id", id);
        params.put("pn", String.valueOf(page));
        params.put("rn", size);
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(BangMusicUrl)
                    .params(params)
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }


    /**
     * 获取歌单的音乐详情
     *
     * @param id   歌单id
     * @param page 页数
     * @param size 长度
     * @return 未解析的数据
     */
    public Single<String> getPlayListMusic(String id, int page, String size) {
        Map<String, String> params = new HashMap<>();
        params.put("type", "songlist");
        params.put("id", id);
        params.put("pn", String.valueOf(page));
        params.put("rn", size);
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(PlayListUrl)
                    .params(params)
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }

    /**
     * 获取搜索的音乐详情
     *
     * @param name 歌曲名称
     * @param page 页数
     * @param size 长度
     * @return 未解析的数据
     */
    public Single<String> getSearchMusic(String name, int page, String size) {
        Map<String, String> params = new HashMap<>();
        params.put("client", "kt");
        params.put("all", name);
        params.put("vipver", "1");
        params.put("ft", "music");
        params.put("encoding", "utf8");
        params.put("rformat", "json");
        params.put("mobi", "1");
        params.put("pn", String.valueOf(page));
        params.put("rn", size);
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(SearchUrl)
                    .params(params)
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }


    /**
     * 获取指定音乐id的播放地址（带音质参数）
     * @param musicId 音乐id
     * @param bitrate 音质码率（如2000、320、128、48）
     * @return 未解析的数据
     */
    public Single<String> getMusicPlayInfo(String musicId, String bitrate) {
        String api = "user=0&android_id=0&prod=kwplayerhd_ar_4.3.0.8&corp=kuwo&vipver=4.3.0.8&source=kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk&notrace=0&type=convert_url2&br=" + bitrate + "&format=flac|mp3|aac&sig=0&priority=bitrate&loginUid=0&network=WIFI&loginSid=0&mode=down&rid=" + musicId;
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(MusicPlayUrl + EncryptMusicToken.encrypt(api))
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }

    /**
     * 获取指定音乐的歌词
     *
     * @param musicId 音乐id
     * @return 未解析的数据
     */
    public Single<String> getMusicLyric(String musicId) {
        Map<String, String> params = new HashMap<>();
        params.put("musicId", musicId);
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(MusicLyricUrl)
                    .params(params)
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }

    /**
     * 获取指定音乐的详情 --- 这里主要是获取歌曲音质
     *
     * @param musicId 歌曲id
     * @return 未解析的数据
     */
    public Single<String> getMusicDetail(String musicId) {
        Map<String, String> params = new HashMap<>();
        params.put("op", "query");
        params.put("action", "play");
        params.put("ids", musicId);
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(MusicDetailUrl)
                    .params(params)
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }

    /**
     * 获取指定音乐id和音质的下载地址
     *
     * @param musicId 音乐id
     * @param bitrate 音质
     * @return 未解析的数据
     */
    public Single<String> getMusicDownloadUrl(String musicId, String bitrate) {
        String api = "user=0&android_id=0&prod=kwplayerhd_ar_4.3.0.8&corp=kuwo&vipver=4.3.0.8&source=kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk&p2p=1&notrace=0&type=convert_url2&br=" + bitrate + "&format=flac|mp3|aac&rid=" + musicId + "&priority=bitrate&loginUid=0&network=WIFI&loginSid=0&mode=down";
        return Single.create(emitter -> {
            HttpRequest.newRequest()
                    .url(MusicPlayUrl + EncryptMusicToken.encrypt(api))
                    .method("GET")
                    .execute(new HttpRequest.Callback() {
                        @Override
                        public void onSuccess(String response) {
                            emitter.onSuccess(response);
                        }

                        @Override
                        public void onError(int statusCode, String message) {
                            emitter.onError(new Exception(message));
                        }

                        @Override
                        public void onFailure(IOException e) {
                            emitter.onError(e);
                        }
                    });
        });
    }


}
